### ISSUE TITLE
Inspect only git tracked files when run under git repos

### ISSUE THREAD

**Issue by ssbarnea** [codespell-project/codespell#1897]
Instead of linting all the folders under current directory or asking the user to maintain a long list of excludes, it would be far better to detect when the tool is run inside a git repository and perform linting only on tracked files: `git ls-files`

This would almost remove the need to maintain exclude lists and provide a very good user experience when running the tool for the first time on any repository.

I already used this approach and testing if inside a git repo is also another very simple command.

**Comment by peternewman** [comment]
This seems to be similar to, but perhaps not quite the same as https://github.com/codespell-project/codespell/issues/1887 . Would you agree?

Can't this be best handled by some external script suggestion, or a pre-commit hook, which just narrows down the list of files to run it against, without tying it entirely to git etc, there are still a few other VCS systems out there.

**Comment by ssbarnea** [comment]
The mentioned one is mostly the same, sorry I did not spot it.

There are other VSC out there but using a wrapper to achieve that does not fix the general UX for the tool itself. Think about calling codespell from other places than pre-commit, like from IDEs (vscode, eclipse). Does this means that every user that want to use codespell would need to spend time managing excludes in different places.

I do prefer to optimize for mos users while not breaking for others. Shortly if not a git repo or git is missing, fallback to old behavior. If someone wants to add support for other VSC, they are welcomed to propose a PR to improve it. 

Considering that I already implemented a git-ls-files in two other python libraries, maybe we should write a very small library that does a smart-glob (uses git ls-files when appropriate but fallbacks to glob). At least it will be easier to reuse in other python projects. One example at https://github.com/ansible-community/ansible-lint/blob/878a90edbce2a985837d26883bff636ec9c7c3a7/src/ansiblelint/file_utils.py#L203-L233

**Comment by KapJI** [comment]
I agree that this should be `codespell` default behaviour (or enabled with single setting) and users don't need to think of how to configure these excluded paths and cover all possible cases basically duplicating `.gitignore`.

I think the ideal UX is when you install it, run `codespell` and it just works.

**Comment by ilyagr** [comment]
> Instead of linting all the folders under current directory or asking the user to maintain a long list of excludes, it would be far better to detect when the tool is run inside a git repository and perform linting only on tracked files: `git ls-files`

Thanks for mentioning `git ls-files`! This made me realize you can just do `codespell $(git ls-files)`. 

(In order to not duplicate the discussion, let's put replies to this in https://github.com/codespell-project/codespell/issues/1887#issuecomment-2611505091 )

**Comment by mdeweerd** [comment]
I opened #3676 .

It should be functionally ok (or at least: almost ok).

Regarding the test case(s) for this new function: I am open to receive PRs from other volunteers (or add access to my fork) to add those.  I think tests can reuse the tests to check the file filter as they are and apply them to the command with the extra option `--git-only`, and add an extra test to validate that files that are not under git control are excluded.
